### ISSUE TITLE
ESM Support - Challenge

### ISSUE THREAD

**Issue by elliot-huffman** [squidfunk/mkdocs-material#5758]
### Context

The statement in the old issue is that it will hurt browser support. I would like to challenge this.

### Description

According to can I use (the same database you use, a great one)
https://caniuse.com/es6-module, if you enable support for this, you are still targeting 95%+ of all users, up by 3% from your target https://squidfunk.github.io/mkdocs-material/browser-support/ of 92%.

You are also putting your project at risk because security issues that have been identified and fixed on Mermaid.JS are not being implemented on your project:
e.g. https://github.com/knsv/mermaid/issues/847 (this example is where more bypasses were found for the old issue and were patched again to prevent more bypasses.)

### Related links

[#5193](https://github.com/squidfunk/mkdocs-material/issues/5193)

### Use Cases

Keeping Deps Up to date and reduce project complexity by allowing to update more deps to the latest version and use newer tech.

### Before submitting

- [X] I have read and followed the [change request guidelines](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/).
- [X] I have verified that [my idea is a change request and not a bug report](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/#its-not-a-bug-its-a-feature).
- [X] I have ensured that, to the best of my knowledge, [my idea will benefit the entire community](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/#benefit-for-the-community).
- [X] I have included relevant links to [the documentation](https://squidfunk.github.io/mkdocs-material), related [issues](https://github.com/squidfunk/mkdocs-material/issues), and [discussions](https://github.com/squidfunk/mkdocs-material/discussions) to underline the need for my idea.

**Comment by facelessuser** [comment]
The main issue that is preventing the latest versions of Mermaind (that uses ESM modules) is that MkDocs doesn't support them.  Once MkDocs supports them, themes like Material can then add support. See relevant discussion: https://github.com/squidfunk/mkdocs-material/discussions/5708.

**Comment by elliot-huffman** [comment]
correct me if I am wrong but wasn't this fixed in 1.5.0?
https://github.com/mkdocs/mkdocs/issues/3164

**Comment by facelessuser** [comment]
Yes, but there are no releases yet for 1.5.0. And once there is, there will need to be work here to add support.  I'm also not sure of the timeline for getting support into Material either, only that I believe there are plans to get it in sometime after 1.5.0.

**Comment by squidfunk** [comment]
Thanks for reporting. We understand the issue and acknowledge the issue. We're not really blocked by MkDocs 1.5.0, because Material for MkDocs auto-loads Mermaid.js once it is needed. [This is done here](https://github.com/squidfunk/mkdocs-material/blob/master/src/assets/javascripts/components/content/code/mermaid/index.ts). More specifically, here:

https://github.com/squidfunk/mkdocs-material/blob/3e8faa0fdba72454f0297d06c21bbde75baa10a2/src/assets/javascripts/components/content/code/mermaid/index.ts#L70-L74

... and here:

https://github.com/squidfunk/mkdocs-material/blob/3e8faa0fdba72454f0297d06c21bbde75baa10a2/src/assets/javascripts/components/content/code/mermaid/index.ts#L87-L133

We're currently very busy with our current docs restructuring, as well as finishing the release of 9.2.0 and the new search integration. __We'd be very happy for help here and we'd love to accept a PR to help upgrade our Mermaid.js integration to ES modules with a fallback for browsers that don't support it.__ We'll also be happy to give assistance in a PR, if it's necessary. Note that ESM integration seems to be rather different in implementation. We're also okay with dropping some of the version range of [the browsers we currently support](https://squidfunk.github.io/mkdocs-material/browser-support/) if it's tiny enough, but this needs more data and further re-evaluation.

Please note that some users are stuck on older systems and if we can offer a "low-cost" fallback, it's definitely the way we should go. This is at least what Material for MkDocs has done in the last years. So, @elliot-huffman, are you up to it? 😊

**Comment by elliot-huffman** [comment]
@squidfunk, I can try as I am a typescript guy, it won't be fast since my day job is a different TypeScript project.

There is a polyfill available in this module but that would require an additional package for this project to build successfully.
However, 95% of browser market share supports ESM modules natively now, I would argue that a polyfil would not be needed.

Marketshare Reference:
https://caniuse.com/es6-module

You target 93% of the MarketShare currently according to your browser support docs page, so this should be compatible with your current target without needing a polyfill to add support for older browsers from my point of view.

**Comment by squidfunk** [comment]
>  I can try as I am a typescript guy, it won't be fast since my day job is a different TypeScript project.

Your help is very much appreciated!

> There is a polyfill available in this module but that would require an additional package for this project to build successfully.
However, 95% of browser market share supports ESM modules natively now, I would argue that a polyfil would not be needed.

Yes, I agree, no polyfill if possible, as we want to keep the JS we ship as small as possible, only loading what's necessary. But we could keep the old way if it turns out to be necessary if we'd be losing significant support range. To assess that, we'd need to update the numbers to find out whether we can drop Chrome 49-60 for instance.

> You target 93% of the MarketShare currently according to your browser support docs page, so this should be compatible with your current target without needing a polyfill to add support for older browsers from my point of view.

Agree. Note that the 92% in our browser support matrix should be updated, as those numbers were last checked in January 2022, so we're likely past the 95% now. In any case, we need to update the numbers.

**Comment by squidfunk** [comment]
Any progress on this? Would love to release it as part of 9.2.0 in the coming weeks.

**Comment by elliot-huffman** [comment]
I will work on it soon if you are ok with the ESM model.
Is it ok based on the can I use data to use Modules with no polyfill?

**Comment by squidfunk** [comment]
Yes, but with a commonjs fallback for Mermaid.js.

**Comment by kallewesterling** [comment]
Are there any updates on this, @elliot-huffman ? Curious as we're having some other issues that are depending on a PR on this issue :)

(see https://github.com/squidfunk/mkdocs-material/discussions/5859)

**Comment by elliot-huffman** [comment]
Sorry about not catching this, I will get to work on it as soon as my current company product feature is completed. That should be within the next 2 weeks.

**Comment by libukai** [comment]
I've successfully utilized the [mkdocs-mermaid2-plugin](https://github.com/fralau/mkdocs-mermaid2-plugin) to resolve the versioning issue with Mermaid 10+. I hope this information can be of help to you all.

However, it's crucial to note that when configuring the superfence, one needs to modify the function used for the `format` based on the template provided by the author:

```
  - pymdownx.superfences:
      custom_fences:
        - name: mermaid
          class: mermaid
          format: !!python/name:mermaid2.fence_mermaid
```

---

希望这样可以满足您的要求！

**Comment by mochaaP** [comment]
I'd like to take this and migrate the build system to parcel v2, since it looks promising for the features we use:

- [inlining](https://parceljs.org/features/bundle-inlining/)
- [assets minification](https://parceljs.org/features/production/#minification)
- [image optimization](https://parceljs.org/recipes/image/)
- [differential bundling](https://parceljs.org/features/targets/#differential-bundling)
- etc.

We could also take this chance to minimize external deps, remove shimmed polyfills and remove `/material` in the repo by utilizing a hatch pre-build hook (e.g. [this](https://github.com/googlefonts/fontra/blob/e65cd1157a4db0a778c23d996bd454608859214b/scripts/rollup_build_hook.py))

**Comment by squidfunk** [comment]
Thanks for suggesting. However, you're proposing very drastic changes. We're very happy with our current homegrown build system that is based on RxJS. It allows us a custom workflow without fighting with APIs from other tools. Previously we had Webpack, and we removed it for a reason. That being said, the `material` folder must be located in the repository, so the theme is installable from git without a build step. This is a deliberate design decision.

**Comment by mochaaP** [comment]
> We're very happy with our current homegrown build system that is based on RxJS. It allows us a custom workflow without fighting with APIs from other tools.

That is precisely the problem I'm facing. The current build system is hard to customize / plug external steps, and the integrations are not very good.

> the `material` folder must be located in the repository, so the theme is installable from git without a build step. This is a deliberate design decision.

I think a build step is necessary. If you don't want a system-wide build time requirement on Node.js, consider <https://github.com/ekalinin/nodeenv>.
Also, a prebuilt package should be published to PyPI / GitHub Actions artifacts (with VCS commit hash on the version tag, preferably with hatch-vcs).

**Comment by elliot-huffman** [comment]
> Sorry about not catching this, I will get to work on it as soon as my current company product feature is completed. That should be within the next 2 weeks.

holey crap, my time has just slipped away, I am sorry I have not been able to get to this sooner. I am still not quite available to work on this feature. As soon as I get some spare time I will get this implemented. I have not forgotten about this!

**Comment by elliot-huffman** [comment]
> I'd like to take this and migrate the build system to parcel v2, since it looks promising for the features we use:
> 
> * [inlining](https://parceljs.org/features/bundle-inlining/)
> * [assets minification](https://parceljs.org/features/production/#minification)
> * [image optimization](https://parceljs.org/recipes/image/)
> * [differential bundling](https://parceljs.org/features/targets/#differential-bundling)
> * etc.
> 
> We could also take this chance to minimize external deps, remove shimmed polyfills and remove `/material` in the repo by utilizing a hatch pre-build hook (e.g. [this](https://github.com/googlefonts/fontra/blob/e65cd1157a4db0a778c23d996bd454608859214b/scripts/rollup_build_hook.py))

This should be on its own issue. This is off topic.

**Comment by squidfunk** [comment]
Yes, I agree, it's definitely off-topic. Please don't discuss the build system here. It has absolutely nothing to do with adding support for ES modules. [You can open a discussion](https://github.com/squidfunk/mkdocs-material/discussions), if you wish to discuss.

**Comment by mochaaP** [comment]
> Please don't discuss the build system here. It has absolutely nothing to do with adding support for ES modules.

My opinion is that differential bundling is vital for supporting native ES modules without sacrificing legacy browser (e.g. Chrome 49-60, Firefox 53-59, Opera 36-47) support.

**Comment by ofek** [comment]
@mochaaP Please post here instead https://github.com/squidfunk/mkdocs-material/issues/6154

**Comment by foxpluto** [comment]
I follow because I need new mermaid version 10 for fixing some issue in my workflow diagram.

**Comment by squidfunk** [comment]
There's currently an open PR at #6265. I just need some time to review it. If you can't wait, you can use the changes introduced in this PR right now and give feedback here whether it solves your problems ☺️

**Comment by squidfunk** [comment]
Fixed in #6265.

**Comment by squidfunk** [comment]
Released as part of [9.4.13](https://squidfunk.github.io/mkdocs-material/changelog/#9.4.13).
